apple/foundationdb
# Data loss protection when joining new cluster

Fixes #5375.

This PR adds a fail-safe to protect a tlog or storage server from deleting its data when joining a new cluster.

This PR introduces a new concept, the cluster ID. The cluster ID is a unique identifier persisted on tlogs and storage servers and stored in the transaction state store. This allows each cluster to have a single unique identifier.

When a tlog or storage server joins a cluster, it will now first read its cluster ID from disk (if it has persisted data). If it has a cluster ID, it will make sure it matches the cluster ID in the txnStateStore. If they don't match, the tlog/storage server will refuse to join the new cluster. Storage servers will log a Sev40, and tlogs will exclude themselves (and log a warning).

## Testing

Passed 100k correctness tests.

Tested manually:

1. Run `fdbserver` and configure a new database.
2. Run another `fdbserver` with a different cluster file.
3. Kill the first `fdbserver` and change its cluster file to use the second cluster file.
4. Running `fdbcli --exec 'status'` should show one excluded process and one process with errors (`StorageServerBelongsToExistingCluster`), viewable through `status json`.
5. Kill the `fdbserver` process from step 3, and run it with the original cluster file. It should rejoin the original cluster after a few seconds, with all its original data intact.

# Code-Reviewer Section

The general guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `master` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)
